README.md
# tinyjs-plugin-transformable

> The Tiny.js plugin for transform something

## 查看demo

http://tinyjs.net/plugins/tinyjs-plugin-transformable.html#demo

## 引用方法

- 推荐作为依赖使用

  - `npm install tinyjs-plugin-transformable --save`

- 也可以直接引用线上cdn地址，注意要使用最新的版本号，例如：

  - https://gw.alipayobjects.com/os/lib/tinyjs-plugin-transformable/1.0.0/index.js
  - https://gw.alipayobjects.com/os/lib/tinyjs-plugin-transformable/1.0.0/index.debug.js

## 起步
首先当然是要引入，推荐`NPM`方式，当然你也可以使用`CDN`或下载独立版本，先从几个例子入手吧！

##### 1、最简单的例子

引用 Tiny.js 源码
``` html
<script src="https://gw.alipayobjects.com/os/lib/alipay/tiny.js/2.2.6/dist/browser/tiny.js"></script>
```
``` js
import Transformable from 'tinyjs-plugin-transformable';

const app = new Tiny.Application({
  //...
});
const container = new Tiny.Container();
app.run(container);

// 创建一个可 transformable 元素
const sprite = Tiny.Sprite.from('https://gw.alipayobjects.com/as/g/tiny/resources/1.0.0/images/logo.png');
const ta = new Tiny.Transformable(sprite);
container.addChild(ta);

```
##### 2、是否贴合

即操作选框、控件是否贴合跟随被编辑的显示对象，默认是贴合的，你可以设置为不贴合：

``` js
const ta = new Tiny.Transformable(sprite, {
  frame: {
    fitness: false
  }
});
```

##### 3、定制参数一览

``` js
const ta = new Tiny.Transformable(sprite, {
  frame: {
    fitness: false,
    thickness: 3,
    color: 0xff0000,
    lineOpacity: 0.5,
    fill: 0xffffff,
    fillOpacity: 0.1,
  },
  drag: {
    inScreen: true,
  },
  zoom: {
    minScale: 0.5,
    maxScale: 1.5,
  },
  rotation: {
    sprite: Tiny.Sprite.fromImage('https://gw.alipayobjects.com/as/g/tiny/resources/1.0.0/images/heart.png'),
  },
  remove: false,
  flipx: false,
  flipy: false,
});
```

##### 4、激活和休眠

``` js
const ta = new Tiny.Transformable(sprite);

container.addChild(ta);
// 添加后立刻激活为编辑态
ta.activate();
// ta.deactivate() // 休眠编辑态

// 点击其他区域切为休眠
container.setEventEnabled(true);
container.hitArea = new Tiny.Rectangle(0, 0, Tiny.WIN_SIZE.width, Tiny.WIN_SIZE.height);
container.on('pointerdown', Tiny.Transformable.deactivateAll);
```

##### 5、固定 Index

``` js
// 按 ta 的添加次序，固定它的 Index，即点击不提升也不置后
ta.fixedIndex(true);
```

##### 6、设定拖拽范围

```js
// 将 Tiny.Transformable 元素的拖拽范围限制在其父 container 相对范围内
Tiny.Transformable.setDragArea(
  // Tiny.Transformable 元素父容器
  container,
  // 相对 container 限制范围，用典型的 left/top + width/height
  { 
    offsetX: 100, // 相对 container 左上角 X
    offsetY: 100, // 相对 container 左上角 Y
    width: 550, // 拖拽范围 width
    height: 550, // 拖拽范围 height
    strict: true, // true: 严格元素在内部，false: 元素锚点在内部
  },
);

```

##### 7、监听回调

``` js
// 点击 remove 按钮后的回调
ta.on('remove:touchend', function(e) {
  console.log('remove', e);
  this.parent.removeChild(this);
});
```

##### 8、添加自定义控件

``` js
var ta = new Tiny.Transformable(sprite, {
  flipx: false, // 将右上角的位置留出来
  flipy: false,
});

ta.addWidget(
  'custom',
  'https://gw.alipayobjects.com/as/g/tiny/resources/1.0.0/images/heart.png',
  [1, -1], // 放在右上角
  {
    onTouchStart: function(e) {
      e.stopPropagation();
      console.log('custom start');
    },
    onTouchMove: function(e) {},
    onTouchEnd: function(e) {},
    onTouchCancel: function(e) {},
  }
);
```

## 依赖
- `Tiny.js`: [Link](http://tinyjs.net/api)

## API文档

http://tinyjs.net/plugins/tinyjs-plugin-transformable.html#docs


CHANGELOG.md
## 1.0.0

`2022-11-17`

- 增加拖拽范围的限制
- 增加对九宫格等非 Sprite 元素的支持
- 自行处理 sprite 加载
- 优化 anchor/pivot 的适配
- bugfix: Widget 使用 renderable 导致不显示状态下还是能点击的问题
- 增加 activate deactivate 事件

## 0.1.0

`2020-02-19`

- 代码重构，更具结构化和扩展性
- 优化缩放的实现，更接近原生操作
- 增加 fitness 贴合设计，你可以选择是否贴合来进行旋转操作
- 完善 Demo 和 API 文档

## 0.0.6

`2020-01-11`

- 修复操作超出画布区域，再操作的异常问题

## 0.0.5

`2020-01-08`

- 增加 `fixedIndex` 方法，可以固定元素的层级

## 0.0.4

`2020-01-07`

- 优化 drag 和 activate 的逻辑，避免不可拖拽时也没法激活的合理性

## 0.0.3

`2020-01-07`

- 增加 `FlipY` 控件
- 修改事件名称，更易于理解和容错
- 修复激活和未激活的 bug

## 0.0.1

`2020-01-06`

- 初次发布


src/Transformable.js
import * as Tiny from '@alipay/tiny.js';
import { DEFAULT_FRAME } from './constants';
import { Draggable, Frame, Zoom, Rotate, Remove, FlipX, FlipY, Widget } from './widget';

/**
 * @example
 * var ta = new Tiny.Transformable(sprite);
 * ta.on('remove:touchend', function(e) {
 *   console.log('remove', e);
 *   this.parent.removeChild(this);
 * });
 *
 * @class
 */
class Transformable extends Tiny.Container {
  /**
   *
   * @param {Tiny.Sprite} target
   * @param {object} opts
   * @param {object} [opts.frame={}]
   * @param {boolean} [opts.frame.fitness=true] - 操作是否贴合
   * @param {number} [opts.frame.thickness=1] - 选框线条粗细
   * @param {number} [opts.frame.color=0xffffff] - 选框线条颜色
   * @param {number} [opts.frame.lineOpacity=1] - 选框线条透明度
   * @param {number} [opts.frame.fill=0] - 选框填充颜色值
   * @param {number} [opts.frame.fillOpacity=0] - 选框填充透明度
   * @param {boolean|object} [opts.drag={}] - 是否可拖拽
   * @param {boolean} opts.drag.inScreen - 是否可移出屏幕外
   * @param {boolean|object} [opts.zoom={}] - 是否可缩放
   * @param {string|Tiny.Texture|Tiny.Sprite} [opts.zoom.sprite=ICONS['zoom'][0]] - 缩放控件的显示对象
   * @param {number} [opts.zoom.minScale=0.5] - 最小缩放值
   * @param {number} [opts.zoom.maxScale=1.5] - 最大缩放值
   * @param {boolean|object} [opts.rotation={}] - 是否可旋转
   * @param {string|Tiny.Texture|Tiny.Sprite} [opts.rotation.sprite=ICONS['rotation'][0]] - 旋转控件的显示对象
   * @param {boolean|object} [opts.remove={}] - 是否显示移除控件
   * @param {string|Tiny.Texture|Tiny.Sprite} [opts.remove.sprite=ICONS['remove'][0]] - 移除控件的显示对象
   * @param {boolean|object} [opts.flipx={}] - 是否可横向翻转
   * @param {string|Tiny.Texture|Tiny.Sprite} [opts.flipx.sprite=ICONS['flipx'][0]] - 横向翻转控件的显示对象
   * @param {boolean|object} [opts.flipy={}] - 是否可纵向翻转
   * @param {string|Tiny.Texture|Tiny.Sprite} [opts.flipy.sprite=ICONS['flipy'][0]] - 纵向翻转控件的显示对象
   */
  constructor(target, opts = {}) {
    super();

    const { frame = {}, drag = {}, zoom = {}, rotation = {}, remove = {}, flipx = {}, flipy = {} } = opts;
    this.$fixedIndex = false;
    this.$icons = {};

    /**
     * @type {Tiny.Sprite}
     */
    this.target = target;

    /**
     * 被编辑显示对象的容器
     *
     * @type {Tiny.Container}
     */
    this.spriteContainer = new Tiny.Container();

    /**
     * 控件的容器
     *
     * @type {Tiny.Container}
     */
    this.widgetContainer = new Tiny.Container();

    /**
     * texture 加载后的初始化
     */
    const textureLoadedFn = () => {
      const _width = target.width;
      const _height = target.height;
      const { width, height } = target.getBounds();

      /**
       * 被编辑显示对象的宽
       *
       * @type {number}
       */
      this.targetWidth = Math.max(_width, width);
      /**
       * 被编辑显示对象的高
       *
       * @type {number}
       */
      this.targetHeight = Math.max(_height, height);
      this.deactivate();
      this.addChild(this.spriteContainer);
      this.addChild(this.widgetContainer);
      this.spriteContainer.addChild(target);
      this.widgetContainer.addChild(Frame.getInstance({ width: this.targetWidth, height: this.targetHeight, ...DEFAULT_FRAME, ...frame }));

      // transformable 元素锚点在中心
      if (target.anchor) {
        target.anchor.set(0.5);
      } else {
        this.spriteContainer.pivot.set(this.targetWidth / 2, this.targetHeight / 2);
      }

      new Draggable(this, drag); // eslint-disable-line

      if (zoom) {
        this.widgetContainer.addChild(Zoom.getInstance(this, { ...{ minScale: 0.5, maxScale: 1.5 }, ...zoom }));
      }

      if (rotation) {
        this.widgetContainer.addChild(Rotate.getInstance(this, { ...rotation, fitness: ({ ...DEFAULT_FRAME, ...frame }).fitness }));
      }

      if (remove) {
        this.widgetContainer.addChild(Remove.getInstance(this, { ...remove }));
      }

      if (flipx) {
        this.widgetContainer.addChild(FlipX.getInstance(this, { ...flipx }));
      }

      if (!flipx && flipy) {
        this.widgetContainer.addChild(FlipY.getInstance(this, { ...flipy }));
      }

      Transformable.instancesPoll.push(this);

      this.on('removed', () => {
        Tiny.arrayRemoveObject(Transformable.instancesPoll, this);
      });
      this.on('added', () => {
        const p = this.parent;
        const cancelHandler = (e) => {
          for (const key in this.$icons) {
            this.$icons[key].activated = false;
          }
        };

        p.on('pointerup', cancelHandler);
        p.on('pointerout', cancelHandler);
        p.on('pointercancel', cancelHandler);
        p.on('pointerupoutside', cancelHandler);
      });
    };

    /**
     * 等 texture 加载成功后初始化
     */
    if (target.isSprite && target.texture.baseTexture.resource) {
      const r = target.texture.baseTexture.resource;
      const loaded = r.load();
      loaded.then(data => {
        textureLoadedFn();
        // console.log(target.width, target.height); // texture 完成加载
      }).catch(error => {
        console.error('sprite image load error: ', error.path); // image 加载失败
      });
    } else {
      textureLoadedFn();
    }

    /**
     * Fired when remove touchend.
     *
     * @event Tiny.Transformable#remove:touchend
     */
    /**
     * Fired when flipx touchend.
     *
     * @event Tiny.Transformable#flipx:touchend
     */
    /**
     * Fired when flipy touchend.
     *
     * @event Tiny.Transformable#flipy:touchend
     */
  }

  /**
   * 激活编辑态
   */
  activate() {
    if (!this.widgetContainer.visible) {
      Transformable.deactivateAll();
      this.widgetContainer.visible = true;

      if (!this.$fixedIndex) {
        this.parent.setChildIndex(this, this.parent.children.length - 1);
      }

      for (const key in this.$icons) {
        this.$icons[key].setEventEnabled(true);
      }

      this.emit('activate');
    }
  }

  /**
   * 休眠编辑态
   */
  deactivate() {
    if (this.widgetContainer.visible) {
      this.widgetContainer.visible = false;
      for (const key in this.$icons) {
        this.$icons[key].setEventEnabled(false);
      }

      this.emit('deactivate');
    }
  }

  /**
   * 是否按添加的顺序固定 index
   *
   * @param {boolean} fixed
   */
  fixedIndex(fixed) {
    this.$fixedIndex = fixed;
  }

  /**
   * 添加自定义控件
   *
   * @param {string} name - 名称，如：skew
   * @param {string|Tiny.Texture|Tiny.Sprite} image - 控件的显示对象
   * @param {number[]} [pos=[1,1]] - 位置（取值区间[[-1, 1], [-1, 1]]），如：[1,1]
   * @param {object} [listeners={}] - 回调
   * @param {function} [listeners.onTouchStart]
   * @param {function} [listeners.onTouchMove]
   * @param {function} [listeners.onTouchEnd]
   * @param {function} [listeners.onTouchCancel]
   */
  addWidget(name, image, pos = [1, 1], listeners = {}) {
    this.widgetContainer.addChild(Widget.getInstance(this, {
      name,
      sprite: image,
      pos,
      listeners,
    }));
  }
}

/**
 * 设定 container 范围
 */
Transformable.setDragArea = Draggable.setDragArea;

/**
 * 当前 transformable 元素
 */
Transformable.instancesPoll = [];

/**
 * 使所有控件都休眠
 *
 * @static
 */
Transformable.deactivateAll = () => {
  Transformable.instancesPoll.forEach(item => {
    item.deactivate();
  });
};

export { Transformable };
